Missing function bodies for MIPS builds

Hello,

I'm using the gorgonia library on MIPS and I get the following errors:

```
../../pkg/mod/github.com/chewxy/math32@v1.10.1/exp.go:3:6: missing function body
../../pkg/mod/github.com/chewxy/math32@v1.10.1/exp.go:57:6: missing function body
../../pkg/mod/github.com/chewxy/math32@v1.10.1/log.go:76:6: missing function body
../../pkg/mod/github.com/chewxy/math32@v1.10.1/remainder.go:33:6: missing function body
../../pkg/mod/github.com/chewxy/math32@v1.10.1/sqrt.go:3:6: missing function body
```

I'm trying to build for mips32, mips32le, mips64, and mips64le and I get the same error. Please advise.

Thanks!

--- a/_exp_arm64.s
+++ b/_exp_arm64.s
@@ -23,7 +23,7 @@
 // This is an assembly implementation of the method used for function Exp in file exp.go.
 //
 // func archExp(x float32) float32
-TEXT ·archExp(SB),$0-16
+TEXT ·archExp(SB),$0-12
 	FMOVS	x+0(FP), F0	// F0 = x
 	FCMPS	F0, F0
 	BNE	isNaN		// x = NaN, return NaN
@@ -109,7 +109,7 @@ overflow:
 // This is an assembly implementation of the method used for function Exp2 in file exp.go.
 //
 // func archExp2(x float32) float32
-TEXT ·archExp2(SB),$0-16
+TEXT ·archExp2(SB),$0-12 // Is this correct?
 	FMOVS	x+0(FP), F0	// F0 = x
 	FCMPS	F0, F0
 	BNE	isNaN		// x = NaN, return NaN

--- a/_stubs_risc64.s
+++ b/_stubs_risc64.s
@@ -1,6 +1,4 @@
-// +build riscv
-
-#include "textflag.h"
+ #include "textflag.h"
 
 // func archExp(x float32) float32
 TEXT ·archExp(SB),NOSPLIT,$0

--- a/exp.go
+++ b/exp.go
@@ -32,7 +32,6 @@ func exp(x float32) float32 {
 	case x > Overflow:
 		return Inf(1)
 	case x < Underflow:
-		return 0
 		// case hx > LogMax:
 		// 	return Inf(1)
 		// case x < 0 && hx > LogMin:

--- a/exp2_asm.go
+++ b/exp2_asm.go
@@ -2,9 +2,9 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !noasm && (arm64 || amd64 || 386 || arm || ppc64le || risc64 || wasm)
+//go:build !noasm && (arm64 || amd64 || 386 || arm || ppc64le || wasm)
 // +build !noasm
-// +build arm64 amd64 386 arm ppc64le risc64 wasm
+// +build arm64 amd64 386 arm ppc64le wasm
 
 package math32
 

--- a/exp2_asm.go
+++ b/exp2_asm.go
@@ -2,9 +2,9 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !noasm && (arm64 || amd64 || 386 || arm || ppc64le || wasm)
+//go:build !noasm && (amd64 || 386 || arm || ppc64le || wasm)
 // +build !noasm
-// +build arm64 amd64 386 arm ppc64le wasm
+// +build amd64 386 arm ppc64le wasm
 
 package math32
 

--- a/exp2_noasm.go
+++ b/exp2_noasm.go
@@ -2,8 +2,8 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build noasm || !(arm64 || amd64 || 386 || arm || ppc64le || risc64 || wasm)
-// +build noasm !arm64,!amd64,!386,!arm,!ppc64le,!risc64,!wasm
+//go:build noasm || !(arm64 || amd64 || 386 || arm || ppc64le || wasm)
+// +build noasm !arm64,!amd64,!386,!arm,!ppc64le,!wasm
 
 package math32
 

--- a/exp2_noasm.go
+++ b/exp2_noasm.go
@@ -2,8 +2,8 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build noasm || !(arm64 || amd64 || 386 || arm || ppc64le || wasm)
-// +build noasm !arm64,!amd64,!386,!arm,!ppc64le,!wasm
+//go:build noasm || !(amd64 || 386 || arm || ppc64le || wasm)
+// +build noasm !amd64,!386,!arm,!ppc64le,!wasm
 
 package math32
 

--- a/exp_asm.go
+++ b/exp_asm.go
@@ -2,9 +2,9 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !noasm && (arm64 || amd64 || 386 || arm || ppc64le || risc64 || wasm || s390x)
+//go:build !noasm && (amd64 || 386 || arm || ppc64le || wasm || s390x)
 // +build !noasm
-// +build arm64 amd64 386 arm ppc64le risc64 wasm s390x
+// +build amd64 386 arm ppc64le wasm s390x
 
 package math32
 

--- a/exp_noasm.go
+++ b/exp_noasm.go
@@ -2,8 +2,8 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build noasm || !(arm64 || amd64 || 386 || arm || ppc64le || risc64 || wasm || s390x)
-// +build noasm !arm64,!amd64,!386,!arm,!ppc64le,!risc64,!wasm,!s390x
+//go:build noasm || !(amd64 || 386 || arm || ppc64le || wasm || s390x)
+// +build noasm !amd64,!386,!arm,!ppc64le,!wasm,!s390x
 
 package math32
 

--- a/export_test.go
+++ b/export_test.go
@@ -3,6 +3,6 @@ package math32
 // Export internal functions for testing.
 
 // var Exp2Go = exp2
-var SqrtGo = noarchsqrt
+var SqrtGo = sqrt
 var ExpGo = exp
 var HypotGo = hypot

--- a/log_asm.go
+++ b/log_asm.go
@@ -2,9 +2,9 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !noasm && (amd64 || s390x || 386 || arm64 || ppc64le || riscv64 || wasm)
+//go:build !noasm && (amd64 || arm || s390x || 386 || arm64 || ppc64le || wasm)
 // +build !noasm
-// +build amd64 s390x 386 arm64 ppc64le riscv64 wasm
+// +build amd64 arm s390x 386 arm64 ppc64le wasm
 
 package math32
 

--- a/log_noasm.go
+++ b/log_noasm.go
@@ -2,8 +2,8 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build noasm || (!amd64 && !s390x && !386 && !arm64 && !ppc64le && !riscv64 && !wasm)
-// +build noasm !amd64,!s390x,!386,!arm64,!ppc64le,!riscv64,!wasm
+//go:build noasm || (!amd64 && !arm && !s390x && !386 && !arm64 && !ppc64le && !wasm)
+// +build noasm !amd64,!arm,!s390x,!386,!arm64,!ppc64le,!wasm
 
 package math32
 

--- a/math32_test.go
+++ b/math32_test.go
@@ -1,0 +1,38 @@
+package math32
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"testing"
+)
+
+func TestVetMath32(t *testing.T) {
+	linuxArches := []string{"amd64", "arm", "arm64", "s390x", "ppc64le", "riscv64", "386", "mips", "mips64", "mipsle", "mips64le"}
+
+	// Linux architectures
+	for _, GOARCH := range linuxArches {
+		GOOS := "linux"
+		t.Run(fmt.Sprintf("GOOS=%s GOARCH=%s", GOOS, GOARCH), func(t *testing.T) {
+			goVet(t, GOOS, GOARCH)
+		})
+	}
+	// WASM
+	t.Run("GOOS=js GOARCH=wasm", func(t *testing.T) {
+		goVet(t, "js", "wasm")
+	})
+
+	// AIX
+	t.Run("GOOS=aix GOARCH=ppc64", func(t *testing.T) {
+		goVet(t, "aix", "ppc64")
+	})
+}
+
+func goVet(t *testing.T, GOOS, GOARCH string) {
+	cmd := exec.Command("go", "vet", ".")
+	cmd.Env = append(os.Environ(), "GOOS="+GOOS, "GOARCH="+GOARCH)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Error(string(output), err)
+	}
+}

--- a/math32_test.go
+++ b/math32_test.go
@@ -15,22 +15,38 @@ func TestVetMath32(t *testing.T) {
 		GOOS := "linux"
 		t.Run(fmt.Sprintf("GOOS=%s GOARCH=%s", GOOS, GOARCH), func(t *testing.T) {
 			goVet(t, GOOS, GOARCH)
+			goBuildVet(t, GOOS, GOARCH)
 		})
 	}
 	// WASM
 	t.Run("GOOS=js GOARCH=wasm", func(t *testing.T) {
 		goVet(t, "js", "wasm")
+		goBuildVet(t, "js", "wasm")
 	})
 
 	// AIX
 	t.Run("GOOS=aix GOARCH=ppc64", func(t *testing.T) {
 		goVet(t, "aix", "ppc64")
+		goBuildVet(t, "aix", "ppc64")
 	})
 }
 
 func goVet(t *testing.T, GOOS, GOARCH string) {
+	env := append(os.Environ(), "GOOS="+GOOS, "GOARCH="+GOARCH)
 	cmd := exec.Command("go", "vet", ".")
-	cmd.Env = append(os.Environ(), "GOOS="+GOOS, "GOARCH="+GOARCH)
+	cmd.Env = env
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Error(string(output), err)
+	}
+}
+
+func goBuildVet(t *testing.T, GOOS, GOARCH string) {
+	env := append(os.Environ(), "GOOS="+GOOS, "GOARCH="+GOARCH)
+	const buildname = "math32.test"
+	defer os.Remove(buildname)
+	cmd := exec.Command("go", "test", "-c", "-o="+buildname)
+	cmd.Env = env
 	output, err := cmd.CombinedOutput()
 	if err != nil {
 		t.Error(string(output), err)

--- a/remainder_asm.go
+++ b/remainder_asm.go
@@ -1,6 +1,6 @@
-//go:build !noasm && (amd64 || s390x)
+//go:build !noasm && (amd64 || s390x || arm || ppc64le || 386 || wasm)
 // +build !noasm
-// +build amd64 s390x
+// +build amd64 s390x arm ppc64le 386 wasm
 
 package math32
 

--- a/remainder_noasm.go
+++ b/remainder_noasm.go
@@ -1,5 +1,5 @@
-//go:build noasm || (!amd64 && !s390x)
-// +build noasm !amd64,!s390x
+//go:build noasm || (!amd64 && !s390x && !arm && !ppc64le && !386 && !wasm)
+// +build noasm !amd64,!s390x,!arm,!ppc64le,!386,!wasm
 
 package math32
 

--- a/sqrt.go
+++ b/sqrt.go
@@ -1,14 +1,21 @@
 package math32
 
+// Sqrt returns the square root of x.
+
+// Special cases are:
+// 	Sqrt(+Inf) = +Inf
+// 	Sqrt(±0) = ±0
+// 	Sqrt(x < 0) = NaN
+// 	Sqrt(NaN) = NaN
 func Sqrt(x float32) float32 {
 	if haveArchSqrt {
 		return archSqrt(x)
 	}
-	return noarchsqrt(x)
+	return sqrt(x)
 }
 
 // TODO: add assembly for !build noasm
-func noarchsqrt(x float32) float32 {
+func sqrt(x float32) float32 {
 	// special cases
 	switch {
 	case x == 0 || IsNaN(x) || IsInf(x, 1):

--- a/sqrt_asm.go
+++ b/sqrt_asm.go
@@ -2,9 +2,9 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build !noasm && (386 || amd64 || arm64 || arm || ppc64le || s390x || riscv64 || wasm)
+//go:build !noasm && (386 || amd64 || arm64 || arm || ppc64le || s390x || wasm)
 // +build !noasm
-// +build 386 amd64 arm64 arm ppc64le s390x riscv64 wasm
+// +build 386 amd64 arm64 arm ppc64le s390x wasm
 
 package math32
 

--- a/sqrt_noasm.go
+++ b/sqrt_noasm.go
@@ -2,8 +2,8 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-//go:build noasm || (!386 && !amd64 && !arm64 && !arm && !ppc64le && !s390x && !riscv64 && !wasm)
-// +build noasm !386,!amd64,!arm64,!arm,!ppc64le,!s390x,!riscv64,!wasm
+//go:build noasm || (!386 && !amd64 && !arm64 && !arm && !ppc64le && !s390x && !wasm)
+// +build noasm !386,!amd64,!arm64,!arm,!ppc64le,!s390x,!wasm
 
 package math32
 

--- a/stubs_arm64.s
+++ b/stubs_arm64.s
@@ -4,5 +4,6 @@
 TEXT ·archLog(SB),NOSPLIT,$0
 	B ·log(SB)
 
-TEXT ·archRemainder(SB),NOSPLIT,$0
-	B ·remainder(SB)
+// func archRemainder(x float32) float32 // TODO
+// TEXT ·archRemainderTODO(SB),NOSPLIT,$0
+//	B ·remainder(SB)

